Resources/doc/configuration.md
# Configuration

## Set up yout OpenID Connect Provider

You will need a working OpenID Connect Identity Provider in order to use the bundle. For testing purposes, you can use
the [MIT ITC Demo Server](https://mitreid.org). The following examples are based on this public service, but remember
to create a client, write down its ID and secret and to enable the `openid`, `email` and `profile` scopes for it.

Authentication between the RP and OP is done through HTTP Basic authentication, so configure your OP properly.

## Configure the bundle

```yaml
#/app/config/config.yml

syntelix_oidc_rp:
    http_client:                     # Configuration for Buzz
        timeout: 5
        verify_peer: ~
        max_redirects: 5
        proxy: ~
    base_url: http://localhost:8000
    client_id: my_client_id          # OpenID Connect client id given by the OpenId Connect Provider
    client_secret: my_client_secret  # OpenID Connect client secret given by the OpenId Connect Provider
    issuer: https://mitreid.org/     # URL of the OpenID Connect Provider
    endpoints_url:                   # Part of the URL of the OpenID Connect Provider
        authorization: /authorize
        token: /token
        userinfo: /userinfo
        logout: /logout
    display: page                    # How the authentication form will be display to the end user
    scope: openid profile email      # List of the scopes you need
    authentication_ttl: 300          # Maximum age of the authentication
    token_ttl: 300                   # Maximum age for tokenID
    jwk_url: https://mitreid.org/jwk # URL to the JSON Web Key of OpenID Connect Provider
    jwk_cache_ttl: 86400             # Validity periods in second where the JWK store in cache is valid
    enabled_state: true              # Enable the use of the state value. This is useful for mitigate replay attack
    enabled_nonce: true              # Enable the use of the nonce value. This is useful for mitigate replay attack
    enduserinfo_request_method: POST # Define the method (POST, GET) used to request the Enduserinfo Endpoint of the OIDC Provider
    redirect_after_logout: /         # URI or route name used for redirect user after a logout
```

## Configure routing

Load the bundle routes.

```yaml
#/app/config/routing.yml

# Syntelix OpenID Connect Relying Party
oidc_rp:
    resource: "@SyntelixOidcRelyingPartyBundle/Resources/config/routing.yml"
```

## Configure security

It is recommended to set a path for `default_target_path` in order to prevent redirection loops.
You should also set a path for `login_path`.

```yaml
#/app/config/security.yml
security:
    providers:
        OICUserProvider: 
            id: syntelix_oidc_rp.user.provider
    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false
        secured_area:
            pattern: ^/
            anonymous: ~
            openidconnect:
                always_use_default_target_path: false
                default_target_path: /
                login_path: /login_check
                target_path_parameter: ~
                use_referer: ~
                create_users: true                 # Create user if not found
                created_users_roles: ROLE_OIC_USER # Add this role(s) to new User
    access_control:
        - { path: ^/private-page, roles: ROLE_OIC_USER }
        - { path: ^/login$, roles: IS_AUTHENTICATED_ANONYMOUSLY }
```


DependencyInjection/Configuration.php
<?php

namespace Syntelix\Bundle\OidcRelyingPartyBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;

/**
 * This is the class that validates and merges configuration from your app/config files.
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * @param $display
     *
     * @return bool
     */
    public static function isHttpMethodSupported($display)
    {
        $displays = array(
            'POST',
            'GET',
        );

        return in_array($display, $displays);
    }

    /**
     * @param $display
     *
     * @return bool
     */
    public static function isDisplaySupported($display)
    {
        $displays = array(
            'page',
            'popup',
            'touch',
            'wap',
            );

        return in_array($display, $displays);
    }

    /**
     * @param $prompt
     *
     * @return bool
     */
    public static function isPromptSupported($prompt)
    {
        $displays = array(
            'none',
            'login',
            'consent',
            'select_account',
            );

        return in_array($prompt, $displays);
    }

    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('syntelix_oidc_rp');

        $this->addHttpClientConfiguration($rootNode);
        $this->addSignatureConfiguration($rootNode);
        $this->addReplayAttackParadeConfiguration($rootNode);

        $rootNode
            ->children()
                ->scalarNode('base_url')->end()
                // URI where the user is redirected after a logout
                ->scalarNode('redirect_after_logout')->defaultNull()->end()
                ->scalarNode('client_id')->cannotBeEmpty()->end()
                ->scalarNode('client_secret')->cannotBeEmpty()->end()
                // issuer is the URL of the OpenId Connect Provider
                // This is needed for validate response of the OpenId Connect Provider
                ->scalarNode('issuer')->cannotBeEmpty()->end()
                // 'token_ttl' value correspond to the iat
                //   Time at which the JWT was issued. Its value is a
                //   JSON number representing the number of seconds
                //   from 1970-01-01T0:0:0Z as measured in UTC until the date/time.
                // Value is stored in second, default 5 minutes
                ->scalarNode('token_ttl')->defaultValue(300)->end()
                // 'authentication_ttl' value correspond to max_age
                //   Maximum Authentication Age. Specifies the allowable elapsed
                //   time in seconds since the last time the End-User was actively
                //   authenticated by the OP. If the elapsed time is greater than
                //   this value, the OP MUST attempt to actively re-authenticate
                //   the End-User. (The max_age request parameter corresponds to
                //   the OpenID 2.0 PAPE [OpenID.PAPE] max_auth_age request parameter.)
                //   When max_age is used, the ID Token returned MUST include
                //   an auth_time Claim Value.
                // Value is stored in second, default 5 minutes
                ->scalarNode('authentication_ttl')->defaultValue(300)->end()
                // @see http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
                ->scalarNode('ui_locales')->end()

                // Define the method (POST, GET) used to request the Enduserinfo Endpoint of the OIDC Provider
                ->scalarNode('enduserinfo_request_method')
                        ->validate()
                        ->ifTrue(function ($display) {
                            return !Configuration::isHttpMethodSupported($display);
                        })
                        ->thenInvalid('Unknown request mathod "%s".')
                    ->end()
                    ->defaultValue('POST')->end()

                // ASCII string value that specifies how the Authorization Server
                // displays the authentication and consent user interface pages
                // to the End-User. The defined values are:
                // @see http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
                ->scalarNode('display')
                    ->validate()
                        ->ifTrue(function ($display) {
                            return !Configuration::isDisplaySupported($display);
                        })
                        ->thenInvalid('Unknown display type "%s".')
                    ->end()
                ->end()

                ->scalarNode('prompt')
                    ->validate()
                        ->ifTrue(function ($display) {
                            return !Configuration::isPromptSupported($display);
                        })
                        ->thenInvalid('Unknown prompt type "%s".')
                    ->end()
                ->end()

                ->scalarNode('scope')
                    ->validate()
                        ->ifTrue(function ($v) {
                            return empty($v);
                        })
                        ->thenUnset()
                    ->end()
                ->end()
            ->end()

            ->children()
                // Endpoints URL are the part of the URL after the OpenId Connect Provider URL
                // If OpenId Connect Provider URL is https://www.myoicop.com/iocp
                // The authorization endpoint configuration is just /authorization
                ->arrayNode('endpoints_url')
                    ->isRequired()
                        ->children()
                            ->scalarNode('authorization')
                                ->validate()
                                    ->ifTrue(function ($v) {
                                        return empty($v);
                                    })
                                    ->thenUnset()
                                ->end()
                            ->end()
                            ->scalarNode('token')
                                ->validate()
                                    ->ifTrue(function ($v) {
                                        return empty($v);
                                    })
                                    ->thenUnset()
                                ->end()
                            ->end()
                            ->scalarNode('userinfo')
                                ->validate()
                                    ->ifTrue(function ($v) {
                                        return empty($v);
                                    })
                                    ->thenUnset()
                                ->end()
                            ->end()
                            ->scalarNode('logout')
                                ->validate()
                                    ->ifTrue(function ($v) {
                                        return empty($v);
                                    })
                                    ->thenUnset()
                                ->end()
                            ->end()
                        ->end()
                ->end()
            ->end()
        ;

        return $treeBuilder;
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addHttpClientConfiguration(ArrayNodeDefinition $node)
    {
        $node
            ->children()
                ->arrayNode('http_client')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('timeout')->defaultValue(5)->cannotBeEmpty()->end()
                        ->booleanNode('verify_peer')->defaultTrue()->end()
                        ->scalarNode('max_redirects')->defaultValue(5)->cannotBeEmpty()->end()
                        ->booleanNode('ignore_errors')->defaultTrue()->end()
                        ->scalarNode('proxy')->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addSignatureConfiguration(ArrayNodeDefinition $node)
    {
        $node
            ->children()
                // URL to the Json Web Key
                ->scalarNode('jwk_url')->defaultNull()->end()
                // Validity periods in second where the JWK is valid
                ->scalarNode('jwk_cache_ttl')->defaultValue(86400)->end()
                // @see http://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata
                ->scalarNode('userinfo_signed_response_alg')->defaultNull()->end()
                ->scalarNode('id_token_signed_response_alg')->defaultNull()->end()
            ->end()
        ;
    }

    /**
     * @param ArrayNodeDefinition $node
     */
    private function addReplayAttackParadeConfiguration(ArrayNodeDefinition $node)
    {
        $node
            ->children()
                ->scalarNode('enabled_state')->defaultTrue()->end()
                ->scalarNode('enabled_nonce')->defaultTrue()->end()
            ->end()
        ;
    }
}
